### PR TITLE
feat: add Binance support

### PR DESCRIPTION
**Pull request by adamant-al**
- Add Binance Support
- Resolve a minor bug with Ladder
- Remove default p2pb2b trade config
- Add optional `loadMarket` param to exchange's API initialization

### CODE REVIEW

File: trade/trader_binance.js [hunk 1]
```diff
@@ -0,0 +1,727 @@
...
+
+    /**
+     * Get info for a specific market
+     * @param pair In readable format as BTC/USDT or in Binance format as BTCUSDT
+     * @returns {Promise<*>|*}
+     */
+    marketInfo(pair) {
+      if (pair?.includes('/')) {
+        return getMarkets(pair);
+      } else {
```
Comment (martiliones) [inline trade/trader_binance.js:120]: Unnecessary `else`

File: trade/trader_binance.js [hunk 2]
```diff
@@ -0,0 +1,727 @@
...
+     */
+    getBalances(nonzero = true) {
+      const paramString = `nonzero: ${nonzero}`;
+
+      return new Promise((resolve) => {
+        binanceApiClient.getBalances().then((data) => {
+          try {
+            let result = [];
+
+            data.balances.forEach((crypto) => {
```
Comment (martiliones) [inline trade/trader_binance.js:162]: use `map`
```js
let result = data.balances.map((crypto) => ({
  code: crypto.asset,
  free: +crypto.free,
  freezed: +crypto.locked,
  total: +crypto.free + +crypto.locked,
}));
```

File: trade/trader_binance.js [hunk 3]
```diff
@@ -0,0 +1,727 @@
...
+    },
+
+    /**
+     * List of all account open orders
+     * @param {String} pair In classic format as BTC/USDT
+     * @returns {Promise<*[]|undefined>}
+     */
+    async getOpenOrders(pair) {
+      const paramString = `pair: ${pair}`;
+      const pair_ = formatPairName(pair);
```
Comment (martiliones) [inline trade/trader_binance.js:194]: confusing variable name

File: trade/trader_binance.js [hunk 4]
```diff
@@ -0,0 +1,727 @@
...
+     * Cancel all order on specific pair
+     * @param pair In classic format as BTC/USDT
+     * @returns {Promise<unknown>}
+     */
+    cancelAllOrders(pair) {
+      const paramString = `pair: ${pair}`;
+      const pair_ = formatPairName(pair);
+
+      return new Promise((resolve, reject) => {
+        binanceApiClient.cancelAllOrders(pair_.pairPlain).then(function(data) {
```
Comment (martiliones) [inline trade/trader_binance.js:353]: use arrow function as you did below

File: trade/trader_binance.js [hunk 5]
```diff
@@ -0,0 +1,727 @@
...
+
+            book.asks.forEach((crypto) => {
+              result.asks.push({
+                amount: +crypto[1],
+                price: +crypto[0],
+                count: 1,
+                type: 'ask-sell-right',
+              });
+            });
+            result.asks.sort(function(a, b) {
```
Comment (martiliones) [inline trade/trader_binance.js:545]: use arrow function

**Review by martiliones — CHANGES_REQUESTED** [review]

**Comment by martiliones** [comment]
I think we should change the structure of the project to be more clear until we have implemented many more exchanges.

From that:

```
.
└── trade
    ├── api
    │   ├── bar_api.js
    │   └── foo_api.js
    ├── settings
    │   ├── tradeParams_bar.js
    │   └── tradeParams_foo.js
    ├── trader_bar.js
    └── trader_foo.js
```

to something like this:

```
.
└── trade
    ├── bar
    │   ├── api.js
    │   ├── index.js
    │   └── settings.js
    ├── foo
    │   ├── api.js
    │   ├── index.js
    │   └── settings.js
    └── shared.js
```

Thus, it should be easier to work on a single exchange and to access shared files, such as utilities

**Comment by adamant-al** [comment]
@martiliones Make a PR to `feature/binance` covering these changes.

File: trade/trader_binance.js [hunk 6]
```diff
@@ -151,37 +155,31 @@ module.exports = (apiKey, secretKey, pwd, log, publicOnly = false, loadMarket =
...
-                free: +crypto.free,
-                freezed: +crypto.locked,
-                total: +crypto.free + +crypto.locked,
-              });
-            });
-
-            if (nonzero) {
-              result = result.filter((crypto) => crypto.free || crypto.freezed);
-            }
+      try {
```
Comment (adamant-al) [inline trade/trader_binance.js:161]: For `async` implementation, don't use nested `try-catch`.
See `getOrderDetails()` as a sample.

File: trade/trader_binance.js [hunk 7]
```diff
@@ -151,37 +155,31 @@ module.exports = (apiKey, secretKey, pwd, log, publicOnly = false, loadMarket =
...
-          log.warn(`API request getBalances(${paramString}) of ${utils.getModuleName(module.id)} module failed. ${err}`);
-          resolve(undefined);
-        });
-      });
+
+          return result;
+        } catch (error) {
+          log.warn(`Error while processing getBalances(${paramString}) request: ${error}`);
+        }
+      } catch (error) {
```
Comment (adamant-al) [inline trade/trader_binance.js:180]: If you change `catch (err)` ⟶ `catch (error)`, change it everywhere in the module.
Same for `catch (e)`.

File: trade/api/binance_api.js [hunk 8]
```diff
@@ -36,104 +40,78 @@ module.exports = function() {
...
-  function getParamsString(data) {
-    const params = [];
+        data.binanceErrorInfo = binanceErrorInfo;
 
-    for (const key in data) {
-      const v = data[key];
-      params.push(key + '=' + v);
-    }
+        if (httpCode >= 400 && httpCode <= 409) {
+          const unexpectedErrorCode = data && (error.code >= -1100 || error.code <= -2013) ?
```
Comment (adamant-al) [inline trade/api/binance_api.js:67]: `error.code > -1100 || error.code < -2013`

File: trade/api/binance_api.js [hunk 9]
```diff
@@ -144,34 +122,34 @@ module.exports = function() {
...
-    if (type !== 'post') {
-      url = getUrlWithParams(url, data);
-    }
-
     return new Promise((resolve, reject) => {
       const httpOptions = {
-        url: url,
+        url,
         method: type,
         timeout: 10000,
```
Comment (adamant-al) [inline trade/api/binance_api.js:137]: Try set `timeout: 1` and run `/balances` command.

Expected:
`Request to https://api.binance.com/api/v3/account with data timestamp=1681096444784&signature=49e893465635be06e99058ff6991e97fb421a23ddeaa1d53c5ff1f3ccc48bd89 failed. Unknown error: AxiosError: timeout of 1ms exceeded. Rejecting…`

In fact: `Error while processing response of request to https://api.binance.com/api/v3/account with data timestamp=1681097715028&signature=e076850fa6286e0c5c523700ad34dc1dd2fbcb4e34bc9f7812c9ebaec755c47a: TypeError: Cannot set properties of undefined (setting 'binanceErrorInfo'). Data object I've got: undefined.`

**Comment by adamant-al** [comment]
@martiliones Good work.
Approve the PR.

**Review by joebrock666 — COMMENTED** [review]
I would suggest to add new lines in the end of the files

**Review by joebrock666 — APPROVED** [review]

**Review by Andrey1029 — APPROVED** [review]

**Pull request merged**